pilot/pkg/networking/apigen/apigen_test.go

```diff
@@ -0,0 +1,193 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apigen_test
+
+import (
+	"context"
+	"log"
+	"os"
+	"testing"
+	"time"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	ads "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/proxy/envoy/xds"
+	"istio.io/istio/pkg/config/schema/collections"
+
+	_ "google.golang.org/grpc/xds/experimental" // To install the xds resolvers and balancers.
+)
+
+
+var (
+	istiodDNSAddr = "127.0.0.1:14053"
+	agentDNSAddr  = "127.0.0.1:14054"
+
+	grpcAddr = "127.0.0.1:14056"
+
+	// Address of the Istiod gRPC service, used in tests.
+	istiodSvcAddr = "istiod.istio-system.svc.cluster.local:14056"
+)
+
+func TestGRPC(t *testing.T) {
+	ds := xds.NewXDS()
+
+	sd := ds.DiscoveryServer.MemRegistry
+	sd.AddHTTPService("fortio1.fortio.svc.cluster.local", "10.10.10.1", 8081)
+	//sd.AddEndpoint("fortio1.fortio.svc.cluster.local",
+	//	"http-main", 8081, "127.0.0.1", 15019)
+
+	sd.AddHTTPService("istiod.istio-system.svc.cluster.local", "10.10.10.2", 14056)
+	sd.SetEndpoints("istiod.istio-system.svc.cluster.local", "", []*model.IstioEndpoint{
+		{
+			Address:         "127.0.0.1",
+			EndpointPort:    uint32(14056),
+			ServicePortName: "http-main",
+		},
+	})
+	se := collections.IstioNetworkingV1Alpha3Serviceentries.Resource()
+	store := ds.MemoryConfigStore
+
+	store.Create(model.Config{
+		ConfigMeta: model.ConfigMeta{
+			Type:      se.Kind(),
+			Group:     se.Group(),
+			Version:   se.Version(),
+			Name:      "fortio",
+			Namespace: "fortio",
+		},
+		Spec: &networking.ServiceEntry{
+			Hosts: []string{
+				"fortio.fortio.svc",
+				"fortio.fortio.svc.cluster.local",
+			},
+			Addresses: []string{"1.2.3.4"},
+
+			Ports: []*networking.Port{
+				{Number: 14056, Name: "grpc-insecure", Protocol: "http"},
+			},
+
+			Endpoints: []*networking.WorkloadEntry{
+				{
+					Address: "127.0.0.1",
+					Ports:   map[string]uint32{"grpc-insecure": 8080},
+				},
+			},
+			Location:   networking.ServiceEntry_MESH_EXTERNAL,
+			Resolution: networking.ServiceEntry_STATIC,
+		},
+	})
+
+	env := ds.DiscoveryServer.Env
+	if err := env.PushContext.InitContext(env, env.PushContext, nil); err != nil {
+		t.Fatal(err)
+	}
+	ds.DiscoveryServer.UpdateServiceShards(env.PushContext)
+
+	err := ds.StartGRPC(grpcAddr)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer ds.GRPCListener.Close()
+
+	os.Setenv("GRPC_XDS_BOOTSTRAP", "testdata/xds_bootstrap.json")
+
+	t.Run("gRPC-resolve", func(t *testing.T) {
```

Comment:
a bit confused why we have grpc test in apigen but we also have grpcgen package? maybe further review will answer this though

Reply:
Merge mistake. In the WIP branch I had the tests in the wrong place, when I split them I forgot to update.